tracing-attributes/src/lib.rs

```diff
@@ -835,127 +968,158 @@ mod kw {
+            return inside_funs
+                .into_iter()
+                .find(|(_, fun)| fun.sig.ident == func_name)
+                .map(|(stmt, fun)| {
+                    // if "_self" is present as an argument, we store its type to be able to rewrite "Self" (the
+                    // parameter type) with the type of "_self"
+                    let self_type = fun
+                        .sig
+                        .inputs
+                        .iter()
+                        .map(|arg| {
+                            if let FnArg::Typed(ty) = arg {
+                                if let Pat::Ident(PatIdent { ident, .. }) = &*ty.pat {
+                                    if ident == "_self" {
+                                        let mut ty = &*ty.ty;
+                                        // extract the inner type if the argument is "&self" or "&mut self"
+                                        if let syn::Type::Reference(syn::TypeReference {
+                                            elem,
+                                            ..
+                                        }) = ty
+                                        {
+                                            ty = &*elem;
+                                        }
+                                        if let syn::Type::Path(tp) = ty {
+                                            return Some(tp.clone());
+                                        }
+                                    }
+                                }
+                            }
 
-    // if "_self" is present as an argument, we store its type to be able to rewrite "Self" (the
-    // parameter type) with the type of "_self"
-    let self_type = fun
-        .sig
-        .inputs
-        .iter()
-        .map(|arg| {
-            if let FnArg::Typed(ty) = arg {
-                if let Pat::Ident(PatIdent { ident, .. }) = &*ty.pat {
-                    if ident == "_self" {
-                        let mut ty = &*ty.ty;
-                        // extract the inner type if the argument is "&self" or "&mut self"
-                        if let syn::Type::Reference(syn::TypeReference { elem, .. }) = ty {
-                            ty = &*elem;
-                        }
-                        if let syn::Type::Path(tp) = ty {
-                            return Some(tp.clone());
-                        }
-                    }
-                }
-            }
+                            None
+                        })
+                        .next()
+                        .flatten();
 
-            None
-        })
-        .next();
-    let self_type = match self_type {
-        Some(x) => x,
-        None => None,
-    };
+                    AsyncTraitInfo {
+                        source_stmt: stmt,
+                        kind: AsyncTraitKind::Function(fun),
+                        self_type,
+                    }
+                });
```

Comment:
this giant expression is quite hard to follow, is it possible to factor some of this out into functions or something to make it more readable?

Reply:
This should be a tad better now